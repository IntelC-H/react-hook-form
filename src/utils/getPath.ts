import isPrimitive from './isPrimitive';
import { FieldName } from '../types';

export const getPath = <TFieldValues>(
  rootPath: FieldName<TFieldValues>,
  values: any,
  paths: FieldName<TFieldValues>[] = [],
): FieldName<TFieldValues>[] => {
  for (const property in values) {
    isPrimitive(values[property])
      ? paths.push(`${rootPath}.${property}` as FieldName<TFieldValues>)
      : getPath(
          `${rootPath}.${property}` as FieldName<TFieldValues>,
          values[property],
          paths,
        );
  }

  return paths;
};
