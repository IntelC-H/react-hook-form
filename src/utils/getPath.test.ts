import { getPath } from './getPath';

describe('getPath', () => {
  it('should generate the correct path', () => {
    expect(
      getPath<any>('test', [
        1,
        [1, 2],
        {
          data: 'test',
          kidding: { test: 'data' },
          foo: { bar: {} },
          what: [{ bill: { haha: 'test' } }, [3, 4]],
          one: 1,
          empty: null,
          absent: undefined,
          isAwesome: true,
          answer: Symbol(42),
        },
      ]),
    ).toMatchSnapshot();

    expect(
      getPath<any>('test', {
        test: 1,
        test1: [1, 2],
        test2: {
          data: 'test',
          kidding: { test: 'data' },
          foo: { bar: {} },
          what: [{ bill: { haha: 'test' } }, [3, 4]],
          one: 1,
          empty: null,
          absent: undefined,
          isAwesome: true,
          answer: Symbol(42),
        },
      }),
    ).toMatchSnapshot();
  });
});
