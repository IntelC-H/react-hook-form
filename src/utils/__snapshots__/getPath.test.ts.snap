// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPath should generate the correct path 1`] = `
Array [
  "test.0",
  "test.1.0",
  "test.1.1",
  "test.2.data",
  "test.2.kidding.test",
  "test.2.what.0.bill.haha",
  "test.2.what.1.0",
  "test.2.what.1.1",
  "test.2.one",
  "test.2.empty",
  "test.2.absent",
  "test.2.isAwesome",
  "test.2.answer",
]
`;

exports[`getPath should generate the correct path 2`] = `
Array [
  "test.test",
  "test.test1.0",
  "test.test1.1",
  "test.test2.data",
  "test.test2.kidding.test",
  "test.test2.what.0.bill.haha",
  "test.test2.what.1.0",
  "test.test2.what.1.1",
  "test.test2.one",
  "test.test2.empty",
  "test.test2.absent",
  "test.test2.isAwesome",
  "test.test2.answer",
]
`;
